crates/plugin-engine/src/api/types.rs

```diff
@@ -0,0 +1,85 @@
+use std::{collections::BTreeMap, path::PathBuf};
+
+use crate::api::backend::BackendAPI;
+use lldap_auth::access_control::ValidationResults;
+
+pub enum PluginSource {
+    ScriptFile(PathBuf),
+    ScriptSource(String),
+}
+
+impl PluginSource {
+    pub fn from_path(path: PathBuf) -> Result<Self, String> {
+        match path.try_exists() {
+            Ok(true) => {
+                if PluginSource::is_lua(&path) {
+                    Ok(PluginSource::ScriptFile(path))
+                } else {
+                    Err("Unrecognized file type".to_string())
+                }
+            }
+            Ok(false) => Err("file does not exist".to_string()),
+            Err(e) => Err(e.to_string()),
+        }
+    }
+
+    fn is_lua(path: &PathBuf) -> bool {
+        path.is_file()
+            && path
+                .extension()
+                .filter(|s| s.eq_ignore_ascii_case("lua"))
+                .is_some()
+    }
+}
+
+impl std::fmt::Display for PluginSource {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            PluginSource::ScriptFile(p) => {
+                write!(
+                    f,
+                    "{}",
+                    p.file_name().map(|f| f.to_str()).flatten().unwrap_or("n/a")
+                )
+            }
+            PluginSource::ScriptSource(s) => {
+                write!(f, "<script: '{} [...]'>", &s.as_str()[0..30])
+            }
+        }
+    }
+}
+
+pub struct PluginConfig {
+    pub plugin_source: PluginSource,
+    pub kvscope: Option<String>,
```

Comment:
Why optional? What does it represent?
I thought the scope was (basically) the plugin name, to partition the DB table?

Reply:
This is basicly the config as given from the LLDAP config file, so the kvscope here is meant for explicitly defining a scope for this plugins storage. If it's None, which would be the general case, it will become the name of the plugin when the plugin is loaded. At this point I haven't accessed the registration info from the plugin yet.